src/cargo/ops/tree/graph.rs

```diff
@@ -391,10 +391,20 @@ fn add_pkg(
         let dep_pkg = graph.package_map[&dep_id];
 
         for dep in deps {
-            let dep_features_for = if dep.is_build() || dep_pkg.proc_macro() {
-                FeaturesFor::HostDep
-            } else {
-                features_for
+            let dep_features_for = match dep
+                .artifact()
+                .and_then(|artifact| artifact.target())
+                .and_then(|target| target.to_resolved_compile_target(requested_kind))
+            {
+                // Dependency has a `{ …, target = <triple> }`
+                Some(target) => FeaturesFor::ArtifactDep(target),
```

Comment:
I might have a rotted memory of artifact dependencies. However, even when `target` is present the dependency might still be depended as a normal `lib` (with `lib = true`). The fix doesn't look 100% correct to me for this reason.

Reply:
Okay this looks like an adaption of my own patch... Need to refresh my memory 😅.